potassium/potassium.py

```diff
@@ -132,69 +135,81 @@ def test_client(self):
-                    res.status_code = 500
-                    res.headers['X-Endpoint-Type'] = endpoint.type
-                    return res
-
-        if endpoint.type == "background":
+            try:
+                out = endpoint.func(req)
+                res = make_response(out.json)
+                res.status_code = out.status
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            except:
+                tb_str = traceback.format_exc()
+                print(colored(tb_str, "red"))
+                res = make_response(tb_str)
+                res.status_code = 500
+                res.headers['X-Endpoint-Type'] = endpoint.type
+            self._gpu_lock.release()
+        elif endpoint.type == "background":
             req = Request(
                 json=flask_request.get_json()
             )
 
             # run as threaded task
             def task(endpoint, lock, req):
-                with lock:
-                    try:
-                        endpoint.func(req)
-                    except Exception as e:
-                        # do any cleanup before re-raising user error
-                        self._write_event_chan(True)
-                        raise e
-                self._write_event_chan(True)
-
-            thread = Thread(target=task, args=(endpoint, self._lock, req))
+                try:
+                    endpoint.func(req)
+                except Exception as e:
+                    # do any cleanup before re-raising user error
+                    raise e
+                finally:
+                    self._background_task_cv.notify_all()
+                    # release lock
+                    lock.release()
+
+            thread = Thread(target=task, args=(endpoint, self._gpu_lock, req))
             thread.start()
 
             # send task start success message
             res = make_response({'started': True})
             res.headers['X-Endpoint-Type'] = endpoint.type
-            return res
+        else:
+            raise InvalidEndpointTypeException()
 
-        raise InvalidEndpointTypeException()
-
-    def _write_event_chan(self, item):
-        try:
-            self._event_chan.put(item, block=False)
-        except Full:
-            pass
+        return res
 
+    # TODO - cover depends on this being called so it should not be private
     def _read_event_chan(self) -> bool:
-        return self._event_chan.get()
+        """
+        _read_event_chan essentially waits for a background task to finish, 
+        and then returns True
+        """
+        with self._gpu_lock:
+            # wait until the background task is done
+            self._background_task_cv.wait()
+        return True
 
-    def _is_working(self):
-        return self._lock.locked()
+    # TODO - we should migrate cover to use this instead of _read_event_chan
+    def wait_for_background_task(self):
+        self._read_event_chan()
+
+    def block_until_gpu_lock_released(self):
```

Comment:
not sure how concurrent lock.aquire() calls prioritize which call site get the lock.

If lock acquisition is based on polling vs a priority or FIFO queue, there's chance in the polling that some other handler/background process grabs the lock and the locked->unlocked transition is never detected by this function

Reply:
we originally had an implementation like this and were worried that if the lock polled, it'd see "locked... locked... locked..." and think that it's the same lock when maybe it was released and then reacquired between polls.

small edge case that we never saw firsthand but worth being aware of. I have not personally verified how lock acquires are implemented.